autopilot/contractor/contractor.go

```diff
@@ -863,12 +869,16 @@ LOOP:
 		}
 
 		// check if we already have a contract with a host on that subnet
-		if shouldFilter && ipFilter.IsRedundantIP(host.NetAddress, host.PublicKey) {
+		if ctx.ShouldFilterRedundantIPs() && filter.IsRedundantIP(host) {
 			continue
 		}
 
+		// form the contract
 		formedContract, proceed, err := c.formContract(ctx, w, host, minInitialContractFunds, maxInitialContractFunds, budget)
-		if err == nil {
+		if err != nil {
+			// remove the host from the filter
+			filter.Remove(host)
```

Comment:
Would a `.Add` make more sense that we only call upon success?

Reply:
I like `Remove` better because generally you want to check+add, only in this small edge case we want to remove the host and it's not even that big of a deal. I'd keep it or even get rid of `.Remove` entirely. I refactored the `contractor` a lot more and then reverted before pushing, this was part of other changes that I did not end up removing because it's technically more correct to remove the host from the filter if formation failed.